### ISSUE TITLE
I hope to know all 57 fields of features, and the fields corresponding to the 57 elements of stds means.

### ISSUE THREAD

**Issue by EvaGalois** [ESkripichnikov/market-making#2]
I used your data processing, and I only got 55 feature data, but you processed 57 data. It seems I still need you to provide some data. . . . Or you can make the readme richer. I changed your code and it stuck with bugs for many days.
can you provide raw data to download? thank you.
I hope to know all 57 fields of features, and the fields corresponding to the 57 elements of stds means.

**Comment by ESkripichnikov** [comment]
> I used your data processing, and I only got 55 feature data, but you processed 57 data. It seems I still need you to provide some data. . . . Or you can make the readme richer. I changed your code and it stuck with bugs for many days. can you provide raw data to download? thank you. I hope to know all 57 fields of features, and the fields corresponding to the 57 elements of stds means.

The two additional features represent Agent State Space features: Inventory Ratio and Total PnL. These features are calculated in real-time as the agent interacts with the environment. For these two features, the mean is set to 0, and the standard deviation is set to 1 to maintain consistency and facilitate model training.

Regarding the raw data, it is not publicly available as it's private to this project due to confidentiality and privacy considerations. Therefore, I'm unable to share the raw data for download. 

If you have any specific questions or need further details on these features or any other aspect of the project, please feel free to ask.

**Comment by EvaGalois** [comment]
I also want to know how many rows of data your features_df takes, and
```python
for i in trange(1, 301):
    a2c.train(strategy, md_train,
              latency=pd.Timedelta(10, 'ms').delta,
              md_latency=pd.Timedelta(10, 'ms').delta,
              count=50,
              train_slice=195_000)
    if i % 500 == 0:
        reward, pnl, trajectory = evaluate(strategy,
                                           md_test,
                                           latency=pd.Timedelta(10, 'ms').delta,
                                           md_latency=pd.Timedelta(10, 'ms').delta)
        wandb.log({
            'val reward': reward,
            'val pnl': pnl,
        })

        torch.save(model.state_dict(), "models/ETH_50step_15_50act_rew_pen.pth")
```
I also want to know how many rows of data your features_df retrieves, and the value principles of count=50 and train_slice=195000 in this code. Are you using a GPU for this training part? How long did it take to train?

In the above code, `.delta` will report an error in my development environment. I replaced it with `.total_seconds()`. I think it is best for me to follow your version.

Another problem is that your action dictionary is divided into 9 actions. I don’t understand the meaning.
```python
self.action_dict = {1: (0, 0), 2: (0, 4), 3: (0, 9),
                            4: (4, 0), 5: (4, 4), 6: (4, 9),
                            7: (9, 0), 8: (9, 4), 9: (9, 9)}
```
What version of python are you using? It would be great if there is a yml file. If not, it doesn’t matter. Thank you for your market-making research on reinforcement learning, which has brought new research directions to the financial field.

I also want to say that you are cool and handsome.

**Comment by EvaGalois** [comment]
This is the feature field details before I save the ESS:
```
 #   Column                Non-Null Count   Dtype         
---  ------                --------------   -----         
 0   receive_ts            880834 non-null  datetime64[ns]
 1   dist_bid_price_0      880834 non-null  float64       
 2   dist_ask_price_0      880834 non-null  float64       
 3   dist_bid_price_1      880834 non-null  float64       
 4   dist_ask_price_1      880834 non-null  float64       
 5   dist_bid_price_2      880834 non-null  float64       
 6   dist_ask_price_2      880834 non-null  float64       
 7   dist_bid_price_3      880834 non-null  float64       
 8   dist_ask_price_3      880834 non-null  float64       
 9   dist_bid_price_4      880834 non-null  float64       
 10  dist_ask_price_4      880834 non-null  float64       
 11  dist_bid_price_5      880834 non-null  float64       
 12  dist_ask_price_5      880834 non-null  float64       
 13  dist_bid_price_6      880834 non-null  float64       
 14  dist_ask_price_6      880834 non-null  float64       
 15  dist_bid_price_7      880834 non-null  float64       
 16  dist_ask_price_7      880834 non-null  float64       
 17  dist_bid_price_8      880834 non-null  float64       
 18  dist_ask_price_8      880834 non-null  float64       
 19  dist_bid_price_9      880834 non-null  float64       
 20  dist_ask_price_9      880834 non-null  float64       
 21  bid_cumul_0           880834 non-null  float64       
 22  ask_cumul_0           880834 non-null  float64       
 23  bid_cumul_1           880834 non-null  float64       
 24  ask_cumul_1           880834 non-null  float64       
 25  bid_cumul_2           880834 non-null  float64       
 26  ask_cumul_2           880834 non-null  float64       
 27  bid_cumul_3           880834 non-null  float64       
 28  ask_cumul_3           880834 non-null  float64       
 29  bid_cumul_4           880834 non-null  float64       
 30  ask_cumul_4           880834 non-null  float64       
 31  bid_cumul_5           880834 non-null  float64       
 32  ask_cumul_5           880834 non-null  float64       
 33  bid_cumul_6           880834 non-null  float64       
 34  ask_cumul_6           880834 non-null  float64       
 35  bid_cumul_7           880834 non-null  float64       
 36  ask_cumul_7           880834 non-null  float64       
 37  bid_cumul_8           880834 non-null  float64       
 38  ask_cumul_8           880834 non-null  float64       
 39  bid_cumul_9           880834 non-null  float64       
 40  ask_cumul_9           880834 non-null  float64       
 41  national_imbalance_0  880834 non-null  float64       
 42  national_imbalance_1  880834 non-null  float64       
 43  national_imbalance_2  880834 non-null  float64       
 44  national_imbalance_3  880834 non-null  float64       
 45  national_imbalance_4  880834 non-null  float64       
 46  national_imbalance_5  880834 non-null  float64       
 47  national_imbalance_6  880834 non-null  float64       
 48  national_imbalance_7  880834 non-null  float64       
 49  national_imbalance_8  880834 non-null  float64       
 50  national_imbalance_9  880834 non-null  float64       
 51  order_flow_imbalance  880834 non-null  float64       
 52  spread                880834 non-null  float64       
 53  CRSI_1m               880834 non-null  float64       
 54  CRSI_5m               880834 non-null  float64       
 55  CRSI_15m              880834 non-null  float64 
 ```

After that:
```python
features_df['ESS'] = features_df.iloc[:, 1:].values.tolist()
features_dict = features_df.set_index('receive_ts')['ESS'].to_dict()

with open('data/features_dict_eth.pickle', 'wb') as f:
    pickle.dump(features_dict, f)
with open('data/features_dict_eth.pickle', 'rb') as f:
    ess_dict = pickle.load(f)
ess_df = pd.DataFrame.from_dict(ess_dict, orient='index').reset_index().rename(columns={'index': 'receive_ts'})

means = ess_df.mean()
stds = ess_df.std()
means.loc['inventory_ratio'] = 0.0
means.loc['tpnl'] = 0.0
stds.loc['inventory_ratio'] = 1.0
stds.loc['tpnl'] = 1.0
```
Am I doing it correctly?

**Comment by ESkripichnikov** [comment]
> I also want to know how many rows of data your features_df retrieves, and the value principles of count=50 and train_slice=195000 in this code.

Regarding the 'count' and 'train_slice' hyperparameters:
'Count' represents the length of a trajectory, which essentially signifies the duration of the agent's interaction with the environment during each training step. It's crucial to strike a balance between a value that is not too small and not too large. This balance ensures that the agent can learn effectively from each trajectory. In my case, the trading strategy places an order every 0.1 seconds (controlled by the 'delay' hyperparameter). Therefore, it makes sense to set the trajectory length based on the frequency of your market data. Typically, I calculate the average length of market data in ticks over a time interval, say 30 seconds, and set this value as the 'count' parameter. However, 'count=50' may seem small, and it could be considered as one of my experiments.

> Are you using a GPU for this training part? How long did it take to train?

The training process primarily involves the interaction with the environment rather than calculating gradients. Consequently, I found it to be more efficient to perform this interaction on a CPU rather than a GPU. Additionally, you can experiment with parallelizing the interaction with the environment. Instead of generating just one trajectory per training step, you could generate, for example, 8 trajectories. This can lead to a smoother and more stable training process.

> Another problem is that your action dictionary is divided into 9 actions. I don’t understand the meaning.

The model outputs a digit from 0 to 9, and these values correspond to specific actions as outlined in a predefined table. For instance, 0 indicates that no action should be taken, while values from 1 to 9 represent different levels at which limit orders should be quoted. For clarity in the code, I use 0 and 0 instead of 1 and 1 as indices in the 'action_dict' to match this convention.
![image](https://github.com/ESkripichnikov/market-making/assets/60779368/16c22f2a-b26d-49b6-81f5-61fec568b3c3)

> What version of python are you using? It would be great if there is a yml file.

I used Python 3.8.13 for this project. I understand the importance of providing a requirements file with all the necessary libraries and their versions. I will definitely include this when I have some free time to prepare it.
 
Your feedback and questions are always welcome, and they help me improve the project and its documentation.

**Comment by ESkripichnikov** [comment]
Regarding the training process: I recall that it took approximately 8-10 hours to complete. For your reference, you can find dynamic training information, including loss and reward metrics, at the following link:
https://wandb.ai/eskripichnikov/MM/reports/RL-in-Market-Making---Vmlldzo1NDUxMTcz

**Comment by ESkripichnikov** [comment]
> Am I doing it correctly?

Looks good. 
You need to get something like this (it's for means, for stds last two digits instead of 0 should be 1):

array([ 4.14099740e+00, -4.14099740e+00,  1.80745342e+01, -1.82680738e+01,
        3.21032290e+01, -3.25206924e+01,  4.51406782e+01, -4.59081977e+01,
        5.80903750e+01, -5.88786379e+01,  7.16807162e+01, -7.21645475e+01,
        8.53400913e+01, -8.55821802e+01,  9.90257197e+01, -9.87863724e+01,
        1.12474618e+02, -1.11663292e+02,  1.25533910e+02, -1.24364046e+02,
        6.95310858e+04,  6.60088915e+04,  8.42230413e+04,  7.92394091e+04,
        9.23483774e+04,  8.66909374e+04,  1.00781549e+05,  9.40509927e+04,
        1.09575530e+05,  1.01706978e+05,  1.18235253e+05,  1.09647435e+05,
        1.26825198e+05,  1.17819090e+05,  1.35392894e+05,  1.26162782e+05,
        1.44048325e+05,  1.34434225e+05,  1.53355785e+05,  1.42678621e+05,
        1.73883767e-03,  2.60108801e-03,  2.10950267e-03,  1.33289979e-03,
        3.52854012e-04, -1.13242746e-03, -2.72523563e-03, -2.84718626e-03,
       -4.00290376e-03, -6.34108624e-03, -9.66782335e-02,  8.28199480e+00,
       -2.46588121e-03,  8.57861564e-03,  3.59268575e-03,  0.00000000e+00,
        0.00000000e+00])

**Comment by EvaGalois** [comment]
> Regarding the training process: I recall that it took approximately 8-10 hours to complete. For your reference, you can find dynamic training information, including loss and reward metrics, at the following link:
https://wandb.ai/eskripichnikov/MM/reports/RL-in-Market-Making---Vmlldzo1NDUxMTcz

I entered this link, and it says: It looks like you’ve landed on a locked or empty page.
If this is the right URL ask the page owner to change the privacy permissions.

my wandbid is `galois-`

**Comment by ESkripichnikov** [comment]
> I entered this link, and it says: It looks like you’ve landed on a locked or empty page.
> If this is the right URL ask the page owner to change the privacy permissions.

Sorry, check this link, please
https://api.wandb.ai/links/eskripichnikov/k2prlao0

**Comment by EvaGalois** [comment]
Thank you.
I would like to know whether you are doing this project for the company or for yourself or society? If it's the latter, I hope to keep in touch with you for a long time. I want to study the reinforcement learning market making model in the future. Glad to meet you.

**Comment by ESkripichnikov** [comment]
I completed this project as part of my master's thesis. It's a pleasure to meet you, and I'm glad to hear about your interest in studying reinforcement learning market-making models in the future. Please feel free to reach out if you have any more questions or need further assistance. You can also connect with me on LinkedIn through the following link:
https://www.linkedin.com/in/eskripichnikov/
